pkg/planner/core/exhaust_physical_plans.go

```diff
@@ -1377,10 +1387,105 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
 	}
 	finalStats := ds.tableStats.ScaleByExpectCnt(rowCount)
 	is.addPushedDownSelection(cop, ds, tmpPath, finalStats)
-	t := cop.convertToRootTask(ds.SCtx())
-	reader := t.p
-	t.p = p.constructInnerByWrapper(wrapper, reader)
-	return t
+	return p.constructIndexJoinInnerSideTask(cop, ds, path, wrapper)
+}
+
+// construct the inner join task by inner child plan tree
+// The Logical include two parts: logicalplan->physicalplan, phsicaplan->task
+// Step1: whether agg can be pushed down to coprocessor
+//
+//	Step1.1: If the agg can be pushded down to coprocessor, we will build a copTask and attach the agg to the copTask
+//	There are two kinds of agg: stream agg and hash agg. Stream agg depends on some conditions, such as the group by cols
+//
+// Step2: build other inner plan node to task
+func (p *LogicalJoin) constructIndexJoinInnerSideTask(dsCopTask *copTask, ds *DataSource, path *util.AccessPath, wrapper *indexJoinInnerChildWrapper) task {
+	if len(wrapper.zippedChildren) == 0 {
+		t := dsCopTask.convertToRootTask(ds.SCtx())
+		return t
+	}
+	la, canPushAggToCop := wrapper.zippedChildren[len(wrapper.zippedChildren)-1].(*LogicalAggregation)
+	if la != nil && la.HasDistinct() {
+		// TODO: remove AllowDistinctAggPushDown after the cost estimation of distinct pushdown is implemented.
+		// If AllowDistinctAggPushDown is set to true, we should not consider RootTask.
+		if !la.SCtx().GetSessionVars().AllowDistinctAggPushDown {
+			canPushAggToCop = false
+		}
+	}
+
+	if canPushAggToCop {
+		// Try stream agg
+		_, preferStream := la.ResetHintIfConflicted()
+		for _, aggFunc := range la.AggFuncs {
+			if aggFunc.Mode == aggregation.FinalMode {
+				preferStream = false
+				break
+			}
+		}
+		// group by a + b is not interested in any order.
+		groupByCols := la.GetGroupByCols()
+		if len(groupByCols) != len(la.GroupByItems) {
+			preferStream = false
+		}
+		if la.HasDistinct() && !la.distinctArgsMeetsProperty() {
+			preferStream = false
+		}
+		// sort items must be the super set of group by items
+		if path != nil && path.Index != nil && ds.tableInfo.GetPartitionInfo() == nil {
+			if len(path.IdxCols) < len(groupByCols) {
+				preferStream = false
+			}
+			sctx := p.SCtx()
+			for i, groupbyCol := range groupByCols {
+				if path.IdxColLens[i] != types.UnspecifiedLength ||
+					!groupbyCol.EqualByExprAndID(sctx.GetExprCtx().GetEvalCtx(), path.IdxCols[i]) {
+					preferStream = false
+				}
+			}
+		} else {
+			preferStream = false
```

Comment:
Can we remove those checks of `preferStream` and just enforce the optimizer to always use `HashAgg`, which can reduce our risks and let the code simpler?

Reply:
This is not better for performance if we directly use ```HashAgg```.